### PR TITLE
[Feature]: Add double flip

### PR DESCRIPTION
**Pull request by yinchimaoliang**
Change double flip op in pipeline, add double-flip in centerpoint detector.

### CODE REVIEW

File: configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus_flip.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+_base_ = ['./centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py']
```
Comment (Tai-Wang) [inline configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus_flip.py:1]: Brackets are not needed. Please also modify other places correspondingly.

File: configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus_flip.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            dict(
+                type='PointsRangeFilter', point_cloud_range=point_cloud_range),
+            dict(
+                type='DefaultFormatBundle3D',
+                class_names=class_names,
+                with_label=False),
+            dict(type='Collect3D', keys=['points'])
+        ],
+        pcd_horizontal_flip=True,
+        pcd_vertical_flip=True)
```
Comment (Tai-Wang) [inline configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus_flip.py:44]: Is it better to place these arguments with flip=True together?

Comment (Tai-Wang) [reply configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus_flip.py:44]: An instruction markdown and comments about how to use test time augmentation here may be helpful.

File: mmdet3d/models/detectors/centerpoint.py [hunk 3]
```diff
@@ -79,3 +81,82 @@ def simple_test_pts(self, x, img_metas, rescale=False):
...
+        merged_bboxes = merge_aug_bboxes_3d(aug_bboxes, scale_img_metas,
+                                            self.pts_bbox_head.test_cfg)
+        return merged_bboxes
+
+    def aug_test(self, points, img_metas, imgs=None, rescale=False):
+        """Test function with augmentaiton."""
+        img_feats, pts_feats = self.extract_feats(points, img_metas, imgs)
+        bbox_list = dict()
+        if pts_feats and self.with_pts_bbox:
+            bbox_pts = self.aug_test_pts(pts_feats, img_metas, rescale)
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/centerpoint.py:160]: bbox_pts -> pts_bbox, make the variable names consistent?

File: mmdet3d/models/detectors/centerpoint.py [hunk 4]
```diff
@@ -79,3 +81,82 @@ def simple_test_pts(self, x, img_metas, rescale=False):
...
         ]
         return bbox_results
+
+    def aug_test_pts(self, feats, img_metas, rescale=False):
+        """Test function of point cloud branch with augmentaiton."""
+        # only support aug_test for one sample
+        outs_list = []
+        for x, img_meta in zip(feats, img_metas):
+            outs = self.pts_bbox_head(x)
+            for task_id, out in enumerate(outs):
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/centerpoint.py:91]: Can add a comment here: merge outputs before decoding bboxes

File: mmdet3d/models/detectors/centerpoint.py [hunk 5]
```diff
@@ -79,3 +81,82 @@ def simple_test_pts(self, x, img_metas, rescale=False):
...
+                        elif key == 'vel':
+                            outs[task_id][0][
+                                key][:, 0,
+                                     ...] = -outs[task_id][0][key][:, 0, ...]
+
+            outs_list.append(outs)
+
+        preds_dicts = dict()
+        scale_img_metas = []
+
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/centerpoint.py:126]: Can add a comment here: concat outputs sharing the same pcd_scale_factor

File: mmdet3d/models/detectors/centerpoint.py [hunk 6]
```diff
@@ -79,3 +81,82 @@ def simple_test_pts(self, x, img_metas, rescale=False):
...
+                scale_img_metas.append(img_meta)
+            else:
+                for task_id, out in enumerate(outs):
+                    for key in out[0].keys():
+                        preds_dicts[pcd_scale_factor][task_id][0][key] += out[
+                            0][key]
+
+        aug_bboxes = []
+        for pcd_scale_factor, preds_dict in preds_dicts.items():
+            for task_id, pred_dict in enumerate(preds_dict):
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/centerpoint.py:140]: Can add a comment here: merge outputs with different flips before decoding bboxes

File: mmdet3d/models/detectors/centerpoint.py [hunk 7]
```diff
@@ -79,3 +81,82 @@ def simple_test_pts(self, x, img_metas, rescale=False):
...
+                for key in pred_dict[0].keys():
+                    preds_dict[task_id][0][key] /= len(outs_list) / len(
+                        preds_dicts.keys())
+            bbox_list = self.pts_bbox_head.get_bboxes(
+                preds_dict, img_metas[0], rescale=rescale)
+            bbox_list = [
+                dict(boxes_3d=bboxes, scores_3d=scores, labels_3d=labels)
+                for bboxes, scores, labels in bbox_list
+            ]
+            aug_bboxes.append(bbox_list[0])
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/centerpoint.py:150]: merge outputs with different scales after decoding bboxes

**Review by Tai-Wang — COMMENTED** [review]
1. Do we need to simplify the configs by inheriting from more high-level bases?
2. Maybe several configs that can validate the function of TTA are enough. Are those all necessary?
3. An instruction of TTA here will be helpful. The readme can be supplemented.

File: mmdet3d/models/dense_heads/centerpoint_head.py [hunk 8]
```diff
@@ -760,9 +760,9 @@ def get_task_detections(self, num_class_with_bg, batch_cls_preds,
                 selected = nms_gpu(
                     boxes_for_nms,
                     top_scores,
-                    thresh=self.test_cfg['nms_iou_threshold'],
-                    pre_maxsize=self.test_cfg['nms_pre_max_size'],
-                    post_max_size=self.test_cfg['nms_post_max_size'])
+                    thresh=self.test_cfg['nms_thr'],
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/centerpoint_head.py:763]: Why do we need to rename them here.

File: mmdet3d/models/detectors/centerpoint.py [hunk 9]
```diff
@@ -79,3 +81,86 @@ def simple_test_pts(self, x, img_metas, rescale=False):
             for bboxes, scores, labels in bbox_list
         ]
         return bbox_results
+
+    def aug_test_pts(self, feats, img_metas, rescale=False):
+        """Test function of point cloud branch with augmentaiton."""
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/centerpoint.py:86]: Add the steps of the function in the docstring thus the users could better understand the implementation of the function. For example
```python
Step 1: map all the feature map back
Step 2: xxxxx
Step 3: xxxxx
```

File: mmdet3d/models/dense_heads/centerpoint_head.py (hunk 8 above)
Comment (yinchimaoliang) [reply mmdet3d/models/dense_heads/centerpoint_head.py:763]: To make it consistant with the parameter names in other models.

File: configs/centerpoint/README.md [hunk 10]
```diff
@@ -33,6 +33,70 @@ We follow the below style to name config files. Contributors are advised to foll
 }
 ```
 
+## Usage
+
+### Test time augmentation
+
+We have supported double-flip and scale augmentation during test time. To use test time augmentation by modifying pipeline and test_cfg in the config.
```
Comment (ZwwWayne) [inline configs/centerpoint/README.md:40]: To use test time augmentation, users need to modify the `test_pipeline` and `test_cfg` in the config.

File: configs/centerpoint/README.md [hunk 11]
```diff
@@ -33,6 +33,70 @@ We follow the below style to name config files. Contributors are advised to foll
...
+For example, we change `centerpoint_0075voxel_second_secfpn_circlenms_4x8_cyclic_20e_nus.py` to the following.
+```python
```
Comment (ZwwWayne) [inline configs/centerpoint/README.md:42]: An empty line above the code block following the markdown lint.

File: configs/centerpoint/README.md [hunk 12]
```diff
@@ -33,6 +33,72 @@ We follow the below style to name config files. Contributors are advised to foll
 }
 ```
 
+## Usage
+
+### Test time augmentation
+
+We have supported double-flip and scale augmentation during test time. To use test time augmentation, users need to modify the
+test_pipeline and test_cfg in the config.
```
Comment (ZwwWayne) [inline configs/centerpoint/README.md:41]:
```python
`test_pipeline` and `test_cfg` in the config.
```

File: configs/centerpoint/README.md [hunk 13]
```diff
@@ -33,6 +33,72 @@ We follow the below style to name config files. Contributors are advised to foll
...
+            dict(type='RandomFlip3D', sync_2d=False),
+            dict(
+                type='PointsRangeFilter', point_cloud_range=point_cloud_range),
+            dict(
+                type='DefaultFormatBundle3D',
+                class_names=class_names,
+                with_label=False),
+            dict(type='Collect3D', keys=['points'])
+        ],
+        pcd_horizontal_flip=True,
```
Comment (Tai-Wang) [inline configs/centerpoint/README.md:93]: Move these two lines just below flip=True

File: mmdet3d/models/detectors/centerpoint.py [hunk 14]
```diff
@@ -79,3 +81,103 @@ def simple_test_pts(self, x, img_metas, rescale=False):
...
+        else:
+            for key in bbox_list[0].keys():
+                bbox_list[0][key] = bbox_list[0][key].to('cpu')
+
+            return bbox_list[0]
+
+    def aug_test(self, points, img_metas, imgs=None, rescale=False):
+        """Test function with augmentaiton."""
+        # import pdb
+        # pdb.set_trace()
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/centerpoint.py:177]: Remove these debug codes

**Review by Tai-Wang — APPROVED** [review]

File: mmdet3d/datasets/pipelines/test_time_aug.py [hunk 15]
```diff
@@ -73,7 +73,7 @@ def __call__(self, results):
                 different scales and flips.
         """
         aug_data = []
-        flip_aug = [False, True] if self.flip else [False]
+        flip_aug = [True] if self.flip else [False]
```
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/test_time_aug.py:76]: Is this modification necessary?

File: mmdet3d/models/detectors/centerpoint.py [hunk 16]
```diff
@@ -79,3 +81,101 @@ def simple_test_pts(self, x, img_metas, rescale=False):
...
+
+    def aug_test_pts(self, feats, img_metas, rescale=False):
+        """Test function of point cloud branch with augmentaiton. The function
+        implementation process is as follows:
+
+        - step 1: map features back for double-flip augmentation.
+        - step 2: merge all features and generate boxes.
+        - step 3: map boxes back for scale augmentation.
+        - step 4: merge results.
+        """
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/centerpoint.py:93]: Use standard docstring like below:
```python
"""Test function of point cloud branch with augmentaiton. 

The function  implementation process is as follows:

        - step 1: map features back for double-flip augmentation.
        - step 2: merge all features and generate boxes.
        - step 3: map boxes back for scale augmentation.
        - step 4: merge results.

Args:

Returns:

```

File: mmdet3d/datasets/pipelines/test_time_aug.py (hunk 15 above)
Comment (yinchimaoliang) [reply mmdet3d/datasets/pipelines/test_time_aug.py:76]: Yes, it will reduce unnecessary scenes when using double flip augmentation during test time.

Comment (ZwwWayne) [reply mmdet3d/datasets/pipelines/test_time_aug.py:76]: We need a comment here as a remark to indicate other developers.

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**